internal/gengapic/paging.go

```diff
@@ -86,82 +87,75 @@ func (g *generator) iterTypeOf(elemField *descriptor.FieldDescriptorProto) (*ite
 	}
 	outMsg, ok := outType.(*descriptor.DescriptorProto)
 	if !ok {
-		return nil, errors.E(nil, "expected %q to be message type, found %T", m.GetOutputType(), outType)
+		return nil, nil, errors.E(nil, "cannot find message type %q, malformed descriptor?", m.GetOutputType())
 	}
 
-	for _, f := range inMsg.Field {
-		if f.GetName() == "page_size" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_INT32 {
-			hasSize = true
-		}
-		if f.GetName() == "page_token" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_STRING {
-			hasToken = true
-		}
-	}
-	for _, f := range outMsg.Field {
-		if f.GetName() == "next_page_token" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_STRING {
-			hasNextToken = true
-		}
-		if f.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
-			elemFields = append(elemFields, f)
+	hasPageToken := false
+	for _, f := range inMsg.GetField() {
+		if (f.GetName() == "page_size" || f.GetName() == "max_results") && f.GetType() == descriptor.FieldDescriptorProto_TYPE_INT32 {
+			if pageSizeField != nil {
+				return nil, nil, errors.E(nil, "found both page_size and max_results fields in message %q", m.GetInputType())
+			}
+			pageSizeField = f
+			continue
 		}
+
+		hasPageToken = hasPageToken || (f.GetName() == "page_token" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_STRING)
 	}
 
-	// If any of the required pagination fields are missing or if the response
-	// has no repeated fields, it is not a fully conforming paginated method.
-	if !hasSize || !hasToken || !hasNextToken || len(elemFields) == 0 {
-		return nil, nil
+	if !hasPageToken || pageSizeField == nil {
+		// Not an error, just not paginated
+		return nil, nil, nil
 	}
 
-	if len(elemFields) > 1 {
-		// ensure that the first repeated field visited has the lowest field number
-		// according to https://aip.dev/4233, and use that one.
-		min := elemFields[0].GetNumber()
-		for _, elem := range elemFields {
-			if elem.GetNumber() < min {
-				return nil, fmt.Errorf("%s: can't pick repeated field in %s based on aip.dev/4233", *m.Name, outType.GetName())
+	hasNextPageToken := false
+	for _, f := range outMsg.GetField() {
+		if f.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
+			if repeatedField != nil {
+				// Repeated fields are tacitly okay as long as the
+				// first listed repeated field has the lowest field number.
+				// In this case, subsequent repeated fields are ignored.
+				// See https://aip.dev/4233 for details.
+				if repeatedField.GetNumber() > f.GetNumber() {
+					return nil, nil, errors.E(nil, "found multiple repeated or map fields in message %q", m.GetOutputType())
+				}
+				// We want the _first_ repeated field to be one paged over.
+				continue
 			}
+			repeatedField = f
 		}
+
+		hasNextPageToken = hasNextPageToken || (f.GetName() == "next_page_token" && f.GetType() == descriptor.FieldDescriptorProto_TYPE_STRING)
```

Comment:
Where do you use `hasNextPageToken`? I suggest something in my other comment below.

Reply:
Good catch. This is a test lacuna.